src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+
+    Statement* s1 = b1->lastStmt();
+
+    GenTree* testTree1 = s1->GetRootNode();
+    noway_assert(testTree1->gtOper == GT_JTRUE);
+
+    // The second and the third block must contain a single statement
+
+    Statement* s2 = b2->firstStmt();
+    if (s2->GetPrevStmt() != s2)
+    {
+        return nullptr;
+    }
+
+    GenTree* testTree2 = s2->GetRootNode();
+
+    if (!optReturnBlock)
+    {
+        noway_assert(testTree2->gtOper == GT_JTRUE);
+    }
+    else
+    {
+        if (testTree2->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        Statement* s3 = b3->firstStmt();
+        if (s3->GetPrevStmt() != s3)
+        {
+            return nullptr;
+        }
+
+        GenTree* testTree3 = s3->GetRootNode();
+        if (testTree3->gtOper != GT_RETURN)
+        {
+            return nullptr;
+        }
+
+        if (!varTypeIsIntegral(testTree2->TypeGet()) || !varTypeIsIntegral(testTree3->TypeGet()))
+        {
+            return nullptr;
+        }
+
+        // The third block is Return with "CNS_INT int 0/1"
+        if (testTree3->AsOp()->gtOp1->gtOper != GT_CNS_INT)
+        {
+            return nullptr;
+        }
+
+        if (testTree3->AsOp()->gtOp1->gtType != TYP_INT)
+        {
+            return nullptr;
+        }
+
+        this->t3 = testTree3;
+    }
+
+    this->t1.testTree = testTree1;
+    this->t2.testTree = testTree2;
+
+    return s1;
+}
+
+//
+// optOptimizeBoolsChkTypeCostCond: Checks if type conditions meet the folding condition, and
+//                                  if cost to fold is not too expensive
+//
+// Arguments:
+//      comp    Instance of Compiler class
+//
+// Return:
+//      True if it meets type conditions and cost conditions.	Else false.
+//
+bool OptBoolsDsc::optOptimizeBoolsChkTypeCostCond(Compiler* comp)
+{
+    noway_assert(t1.compTree->OperIs(GT_EQ, GT_NE) && t1.compTree->AsOp()->gtOp1 == c1);
+    noway_assert(t2.compTree->OperIs(GT_EQ, GT_NE) && t2.compTree->AsOp()->gtOp1 == c2);
+
+    //
+    // Leave out floats where the bit-representation is more complicated
+    // - there are two representations for 0.
+    //
+    if (varTypeIsFloating(c1->TypeGet()) || varTypeIsFloating(c2->TypeGet()))
+    {
+        return false;
+    }
+
+    // Make sure the types involved are of the same sizes
+    if (genTypeSize(c1->TypeGet()) != genTypeSize(c2->TypeGet()))
+    {
+        return false;
+    }
+    if (genTypeSize(t1.compTree->TypeGet()) != genTypeSize(t2.compTree->TypeGet()))
+    {
+        return false;
+    }
+#ifdef TARGET_ARMARCH
+    // Skip the small operand which we cannot encode.
+    if (varTypeIsSmall(c1->TypeGet()))
+        return false;
+#endif
+    // The second condition must not contain side effects
+
+    if (c2->gtFlags & GTF_GLOB_EFFECT)
+    {
+        return false;
+    }
+
+    // The second condition must not be too expensive
+
+    comp->gtPrepareCost(c2);
+
+    if (c2->GetCostEx() > 12)
```

Comment:
I wonder what kind of diffs there are if we eliminate this cost check. That is, is 12 the right number, and why?

Reply:
TODO:
It is from the original code. Will move to future work. It can be handled in the context of PGO.

Reply:
Created https://github.com/dotnet/runtime/issues/55605 and added this item there.